### PR TITLE
[LIBCLOUD-604] Add 'ex_list_subnets' to OpenStack provider

### PR DESCRIPTION
**Pull request by hannanabdul55**
Added ability to list subnets for libcloud OpenStack driver by fetching them from the OpenStack API and parsing them to a list of `OpenStackSubnet` class.  
This has been tested with a local devstack environment setup on my local machine.
Reference to the OpenStack API [here.](http://developer.openstack.org/api-ref-networking-v2.html#listSubnets)

### CODE REVIEW

**Comment by Kami** [comment]
Thanks.

Can you please also add a corresponding test case?

**Comment by hannanabdul55** [comment]
Yeah sure!

**Comment by hannanabdul55** [comment]
So does the Openstack test cases use rackspace driver in the backend? Because in my test case for the above code, I call the `driver.ex_list_subnets` function which throws an error saying this function is undefined in `RackspaceFirstGenNodeDriver` .  
I agree that Rackspace runs on OpenStack, but shouldn't these tests be performed on a running devstack instance?

**Comment by Kami** [comment]
@hannanabdul55 All the tests right right now are unit tests where we use mock provider responses (we don't test against a live environment).

Right now OpenStack tests inherit from Rackspace oned and not vice versa and that's something which we eventually need to clean up.

I'll have a look at what could be causing the issue and let you know (we do some "weird" test inheritance which might be causing the issue).

**Comment by hannanabdul55** [comment]
Yeah sure. I have the unit tests for the same ready anyways. But it breaks the testing, hence, should I withhold the commit for the test cases as of now?

**Comment by Kami** [comment]
@hannanabdul55 Can you gist the tests / diff somewhere? And I will have a look.

**Comment by hannanabdul55** [comment]
I have lost that local repository. I will gist it in a while after recovering it.

File: libcloud/compute/drivers/openstack.py [hunk 1]
```diff
@@ -335,6 +335,34 @@ def ex_hard_reboot_node(self, node):
...
+    def ex_list_subnet(self, network_id=None):
+        """
+        List all subnet resources
+        :param        network_id:   filter for subnets for this network ID
+        :type         network_id:   ``str``
+        :rtype:       ``list`` of :class:``OpenStackSubnet``
+        """
+        uri = '/subnets'
+        resp = self.connection.request(uri, method='GET')
+        if resp.status == httplib.NOT_FOUND:
```
Comment (Kami) [inline libcloud/compute/drivers/openstack.py:347]: I assume this means that this functionality (neutron?) is not available?

In any case, it's probably better to return an empty list.

File: libcloud/compute/drivers/openstack.py [hunk 2]
```diff
@@ -335,6 +335,34 @@ def ex_hard_reboot_node(self, node):
...
+            return None
+        return self._to_subnets(resp.object)
+
+
+class OpenStackSubnet(object):
```
Comment (Kami) [inline libcloud/compute/drivers/openstack.py:352]: To make it easier to eventually switch to #345, can you please make sure this class looks more similar to the one here https://github.com/apache/libcloud/pull/345/files#diff-4c75f73aaa54353918c9fc5a4fc083e5R102 https://github.com/apache/libcloud/pull/345/files#diff-5cc9e28889b2f8491dcdbc2114a213b0R65?

**Comment by Kami** [comment]
Sorry for the delay.

Added some comments.

**Comment by hannanabdul55** [comment]
Yeah sure. I will do this and commit!

**Pull request closed without merging**